File: mahaam-api-py/infra/email.py
import uuid
from twilio.rest import Client
from twilio.base.exceptions import TwilioRestException
from infra import db, configs
from infra.log import Log

client: Client = None


def init():
    global client
    client = Client(configs.data.emailAccountSid, configs.data.emailAuthToken)


def send_otp(email: str) -> tuple[str, Exception]:
    try:
        verification = client.verify.v2.services(configs.data.emailVerificationServiceSid) \
            .verifications \
            .create(to=email, channel='email')
        return verification.sid, None
    except TwilioRestException as e:
        Log.error(str(uuid.UUID(int=0)),
                  f"Error sending OTP to {email}: {str(e)}")
        return "", e


def verify_otp(otp: str, sid: str, email: str) -> tuple[str, Exception]:
    try:
        verification_check = client.verify.v2.services(configs.data.emailVerificationServiceSid) \
            .verification_checks \
            .create(to=email, code=otp, verification_sid=sid)
        return verification_check.status, None
    except TwilioRestException as e:
        Log.info(str(uuid.UUID(int=0)),
                 f"Error verifying OTP for {email}: {str(e)}")
        return "", e

File: mahaam-api-ts/src/infra/email.ts
/* eslint-disable */

import * as twilio from 'twilio';
import config from './config';

const client = twilio(config.emailAccountSid, config.emailAuthToken);

//    function init()  {
//     try {

//       this.verificationServiceSid = Config.emailVerificationServiceSid;
//     } catch (error: unknown) {
//       console.error('Failed to initialize Twilio client:', error);
//       throw error;
//     }
//   }

async function sendOtp(email: string): Promise<string | null> {
  try {
    if (!client || !config.emailVerificationServiceSid) {
      throw new Error('Email service not initialized. Call Email.init() first.');
    }

    const verification = await client.verify.v2.services(config.emailVerificationServiceSid).verifications.create({
      to: email,
      channel: 'email',
    });

    return verification.sid;
  } catch (error: unknown) {
    console.error('Failed to send OTP:', error);
    return null;
  }
}

async function verifyOtp(otp: string, sid: string, email: string): Promise<string | null> {
  try {
    if (!client || !config.emailVerificationServiceSid) {
      throw new Error('Email service not initialized. Call Email.init() first.');
    }

    const verificationCheck = await client.verify.v2.services(config.emailVerificationServiceSid).verificationChecks.create({
      to: email,
      code: otp,
    });

    return verificationCheck.status;
  } catch (error: unknown) {
    console.error('Failed to verify OTP:', error);
    return null;
  }
}

export default {
  sendOtp,
  verifyOtp,
} as const;
